### PR TITLE
Modify sbrk to zero fill allocated space

### PR DESCRIPTION
**Pull request by mikbras**
This PR addresses issue #2373 which reports a security vulnerability resulting from a design decision not to initialize and measure heap pages during enclave loading. Please see the issue for background information.

This PR modifies **oe_sbrk()** to zero-fill memory incrementally as it is allocated. The PR also fixes the unhandled case where the increment parameter is negative.

### CODE REVIEW

File: enclave/core/sbrk.c [hunk 1]
```diff
@@ -1,32 +1,89 @@
...
+        ret = _break;
+        _break += increment;
+
+        if (_break > _max)
         {
-            ptr = _heap_next;
-            _heap_next += increment;
+            /* Zero-initializes newly allocated memory. */
+            memset(_max, 0, (size_t)(_break - _max));
+            _max = _break;
```
Comment (radhikaj) [inline enclave/core/sbrk.c]: Thanks a lot for this PR @mikbras
Would using memset_s be good?

**Comment by mingweishih** [comment]
I'm still not sure we need to initialize memory in sbrk as the use of heap memory usually depends on the behavior of the program (e.g., malloc v.s. calloc). It would be great to firstly check potential attacks rooted from this issue. For example, one case I'm thinking of is that the attacker creates a new enclave, allocates the memory overlapped with the heap of the victim enclave, and reads the leftover content. I'm wondering would this be the case, or are there other potential attack vectors.

**Comment by jhand2** [comment]
Do we (or sgx implementation) clear pages when they are released from EPC? If not that is probably worse that the vulnerabilities this protects against.

I talked with MingWei about this and he has a good point that this isn't really a TEE specific mitigation. Using uninitialized memory is just a general security bug, regardless of whether it runs in SGX. Maybe as OE we should help developers write secure code in general, but we should discuss if that is a goal or not.

We also talked about the perf penalty. We should probably come up with a way to profile this code to determine how large of a penalty we're adding for enclave applications.

**Comment by CodeMonkeyLeet** [comment]
> Do we (or sgx implementation) clear pages when they are released from EPC? If not that is probably worse that the vulnerabilities this protects against.

I'm not quite sure I understand the question here: Currently in SGX, a page is either an EPC page or it is not as allocated based on the BIOS setting; it is not converted to/from the memory pool available to the OS. An EPC page can be added/removed from a specific enclave via EADD/EREMOVE (and eventually EAUG) but I imagine that it's ultimately the responsibility of the EPC manager (i.e. Windows OS or Linux SGX driver) to zero EPC pages reclaimed from an enclave.

Just as it's the case that the SDK doesn't invoke EREMOVE itself today (it relies on the virtual memory manager to do so when the enclave memory region is freed by the host process) it can't guarantee that EPC pages it used are cleaned up if the enclave terminates unexpectedly (or is killed by a malicious host).

**Comment by jhand2** [comment]
> When an enclave is created, its heap pages are set to 0.

Looks like you're right.

@mikbras see the `host/sgx/create.c` function `_add_heap_pages`. It calls `_add_filled_pages` which zeros the page and then adds it.

Can you confirm this understanding? Then I agree, we can close this PR and issue.

**Comment by jhand2** [comment]
I'm closing this PR. Please reopen if you feel it is still required

**Comment by mikbras** [comment]
> > When an enclave is created, its heap pages are set to 0.
> 
> Looks like you're right.
> 
> @mikbras see the `host/sgx/create.c` function `_add_heap_pages`. It calls `_add_filled_pages` which zeros the page and then adds it.
> 
> Can you confirm this understanding? Then I agree, we can close this PR and issue.

Yes, the loader zero-initializes heap pages but a malicious loader may not. An alternative is for **sbrk()** to simply check whether extended regions are zero-initialized.

**Pull request closed without merging**